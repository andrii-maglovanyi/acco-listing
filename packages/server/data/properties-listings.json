{
  "properties": [
    {
      "id": "1",
      "images": ["104_2160.jpg", "174_2160.jpg", "191_2160.jpg"],
      "bedrooms": 2,
      "bathrooms": 1,
      "address": "136  Crescent Avenue, DRUMELZIER",
      "postcode": "ML12 2FS",
      "description": "Are you looking for a nice home in a green, centrally located, child-friendly environment? Then we have good news! This house on the DRUMELZIER is perfectly located and surrounded by greenery.",
      "price": 750.0,
      "expired": false
    },
    {
      "id": "2",
      "images": ["106_3280.jpg", "111_3280.jpg", "154_3280.jpg"],
      "bedrooms": 4,
      "bathrooms": 2,
      "address": "26  Bouverie Road, WESTON",
      "postcode": "CW2 5JG",
      "description": "Here is another nice starter home. Modern kitchen, toilet and bathroom. The house is stylishly finished, completely contemporary. You can go in! An interesting middle house for starters and young families. ",
      "price": 500.0,
      "expired": true
    },
    {
      "id": "3",
      "images": ["218_2160.jpg", "221_2160.jpg", "226_2160.jpg"],
      "bedrooms": 1,
      "bathrooms": 1,
      "address": "76  Hudson St, DUNCTON",
      "postcode": "GU28 4LR",
      "description": "If interested, we are happy to take the time for you! Plan an appointment via our website and we will be happy to show you the home with a maximum of 2 people. If a physical appointment does not work out well, call for digital options!",
      "price": 1100.0,
      "expired": true
    },
    {
      "id": "4",
      "images": ["235_3280.jpg", "251_3280.jpg", "297_3280.jpg"],
      "bedrooms": 7,
      "bathrooms": 3,
      "address": "55  Thornton St, HUTTON BUSCEL",
      "postcode": "YO13 0XR",
      "description": "Living outside but still in the city. This detached villa with a double detached garage is situated in a very beautiful location in the popular villa district of HUTTON BUSCEL. This spacious family house was newly built in 2007 and is situated on a very spacious plot, surrounded by a beautiful and landscaped garden that offers a lot of privacy.",
      "price": 480.0,
      "expired": true
    },
    {
      "id": "5",
      "images": ["319_3280.jpg", "361_3280.jpg", "365_3280.jpg"],
      "bedrooms": 1,
      "bathrooms": 2,
      "address": "52  St Maurices Road, PRESTON UPON THE WEALD",
      "postcode": "TF6 9ZG",
      "description": "What a great house! Very attractive, full of beautiful materials and details and perfectly maintained. Spacious living room, kitchen, cellar, 3 bedrooms and bathroom on the 1st floor, attic and beautifully landscaped backyard on the South East.",
      "price": 945.0,
      "expired": false
    },
    {
      "id": "6",
      "images": ["361_2160.jpg", "395_2160.jpg", "369_2160.jpg"],
      "bedrooms": 2,
      "bathrooms": 2,
      "address": "50  Academy Street, BENBUIE",
      "postcode": "DG3 3XZ",
      "description": "Spacious, ready-to-live-in family house with bright rooms, four spacious bedrooms, modern kitchen, luxury bathroom and a sunny south-facing garden.",
      "price": 475.0,
      "expired": true
    },
    {
      "id": "7",
      "images": ["413_2160.jpg", "427_2160.jpg", "432_2160.jpg"],
      "bedrooms": 4,
      "bathrooms": 1,
      "address": "107  Hendford Hill, MOY",
      "postcode": "PH33 6DW",
      "description": "A surprisingly spacious detached house with garage located on a quiet street on the outskirts of MOY.",
      "price": 800.0,
      "expired": false
    },
    {
      "id": "8",
      "images": ["510_2160.jpg", "517_2160.jpg", "582_2160.jpg"],
      "bedrooms": 2,
      "bathrooms": 3,
      "address": "94  Winchester Rd",
      "postcode": "MICKLETOWN",
      "description": "Are you looking for an energy-efficient corner house, close to shopping center then a visit to this house is definitely worth it! This maintenance-friendly house with garage, storage and garden is located on 224 mÂ².",
      "price": 240.0,
      "expired": true
    }
  ]
}
