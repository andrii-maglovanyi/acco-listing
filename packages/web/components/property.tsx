import { useState } from "react";
import styled from "styled-components";

import { Property as PropertyModel } from "@acco-listing/server/src/properties/property.model";

import { maybePluralize, textTruncate } from "../utils/text-transform";

interface Props extends PropertyModel {
  onChangeStatus: (expired: boolean) => void;
}

const Wrapper = styled.section`
  border-radius: 4px;
  display: flex;
  margin-bottom: 1rem;
  overflow: hidden;

  .expired {
    opacity: 0.5;

    img {
      cursor: default;
    }

    &::after {
      background: white;
      content: "expired";
      font-size: 1.5rem;
      left: 64%;
      padding: 0.5rem 2.2rem;
      position: absolute;
      text-transform: uppercase;
      top: 74%;
      transform: rotate(-45deg);
    }
  }
`;

const Address = styled.div`
  font-size: 0.8rem;
  color: var(--dark-gray-color);
`;

const ImageContainer = styled.div`
  cursor: pointer;
  flex: 1 1 350px;
  position: relative;
  height: 250px;
  width: 350px;
  overflow: hidden;

  img {
    position: absolute;
    height: auto;
    width: 100%;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%) scale(1.1);
    animation: imageLoad ease 0.3s;
  }
`;

const Details = styled.div`
  border: 1px solid var(--gray-color);
  border-left: none;
  border-radius-bottom-right: 4px;
  border-radius-top-right: 4px;
  box-sizing: border-box;
  flex: 1 1 450px;
  flex-direction: column;
  padding: 1rem;
  position: relative;
`;

const Price = styled.div`
  color: var(--purple-color);
  font-weight: bold;

  .guide-price {
    font-size: 0.8rem;
  }

  .price {
    font-size: 1.6rem;
    margin: 0.2rem 0 1rem 0;
  }
`;

const StatusButton = styled.button`
  cursor: pointer;
  position: absolute;
  right: 1rem;
  top: 1rem;
`;

export const Property: React.FC<Props> = (props) => {
  const {
    images,
    bedrooms,
    bathrooms,
    address,
    postcode,
    description,
    price,
    expired,
  } = props;

  const [imageNum, setImageNum] = useState(0);

  const nextImage = (): void => {
    if (!expired) {
      setImageNum(images.length - 1 > imageNum ? imageNum + 1 : 0);
    }
  };

  return (
    <Wrapper data-testid="property">
      <ImageContainer
        onClick={nextImage}
        className={expired ? "expired" : "active"}
      >
        <img
          alt={`${maybePluralize(
            bedrooms,
            "bedroom"
          )} house for sale in ${address}, ${postcode}`}
          src={`/images/${images[imageNum]}`}
        />
      </ImageContainer>
      <Details>
        <Price>
          <div className="guide-price">Guide price</div>
          <p className="price">&pound;{price}</p>
        </Price>
        <p>
          <span className="icon icon-bedroom"></span> {bedrooms}{" "}
          <span className="icon icon-bathroom"></span> {bathrooms}
        </p>
        <div>
          <strong>{maybePluralize(bedrooms, "bed")} house for sale</strong>
        </div>
        <Address>
          {address}, {postcode}
        </Address>
        <p>{textTruncate(description)}</p>
        <StatusButton
          onClick={(): void => props.onChangeStatus(expired)}
          onKeyPress={(): void => props.onChangeStatus(expired)}
          tabIndex={-1}
        >
          {expired ? "Expired" : "Active"}
        </StatusButton>
      </Details>
    </Wrapper>
  );
};
