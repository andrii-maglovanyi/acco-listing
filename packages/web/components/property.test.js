import { cleanup, fireEvent, render } from "@testing-library/react";

import { properties } from "@acco-listing/server/data/properties-listings.json";
import { Property } from "./property.tsx";

afterEach(cleanup);

describe("Property", () => {
  it("Should render the component", () => {
    const container = render(<Property {...properties[1]} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("Should call status change handler", async () => {
    const onClick = jest.fn();
    const { getByText } = render(
      <Property {...properties[0]} onChangeStatus={onClick} />
    );

    const summary = getByText("2 beds house for sale");
    expect(summary).toBeTruthy();

    const button = getByText("Active");
    expect(onClick).not.toHaveBeenCalled();
    fireEvent.click(button);
    expect(onClick).toHaveBeenCalledTimes(1);
    expect(onClick).toHaveBeenCalledWith(false);
  });

  test("Should rotate images on click", () => {
    const { getByAltText } = render(<Property {...properties[0]} />);

    const image = getByAltText(/2 bedrooms house for sale/);
    expect(image.src).toContain("104_2160.jpg");
    fireEvent.click(image.parentElement);
    expect(image.src).toContain("174_2160.jpg");
    fireEvent.click(image.parentElement);
    expect(image.src).toContain("191_2160.jpg");
    fireEvent.click(image.parentElement);
    expect(image.src).toContain("104_2160.jpg");
  });
});
