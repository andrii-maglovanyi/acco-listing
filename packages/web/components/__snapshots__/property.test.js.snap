// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Property Should render the component 1`] = `undefined`;
